src/lib.rs

```diff
@@ -46,329 +43,62 @@
 
 use std::{
     io::{Error, ErrorKind},
-    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, UdpSocket},
+    net::IpAddr,
 };
 
-// Though the module includes `allow(clippy::all)`, that doesn't seem to affect some lints
-#[allow(clippy::semicolon_if_nothing_returned, clippy::struct_field_names)]
-#[cfg(windows)]
-mod win_bindings;
+#[cfg(any(apple, bsd))]
+use bsd::interface_and_mtu_impl;
+#[cfg(target_os = "linux")]
+use linux::interface_and_mtu_impl;
+#[cfg(target_os = "windows")]
+use windows::interface_and_mtu_impl;
 
-/// Prepare a default error result.
-fn default_result<T>() -> Result<(String, T), Error> {
-    Err(Error::new(
-        ErrorKind::NotFound,
-        "Local interface MTU not found",
-    ))
-}
+#[cfg(any(apple, bsd))]
+mod bsd;
 
-#[derive(Debug)]
-pub enum SocketAddrs {
-    Local(IpAddr),
-    Remote(SocketAddr),
-    Both((IpAddr, SocketAddr)),
-}
+#[cfg(target_os = "linux")]
+mod linux;
 
-impl From<&(IpAddr, SocketAddr)> for SocketAddrs {
-    fn from((local, remote): &(IpAddr, SocketAddr)) -> Self {
-        Self::Both((*local, *remote))
-    }
-}
+#[cfg(target_os = "windows")]
+mod windows;
 
-impl From<&(Option<IpAddr>, SocketAddr)> for SocketAddrs {
-    fn from((local, remote): &(Option<IpAddr>, SocketAddr)) -> Self {
-        local.map_or(Self::Remote(*remote), |local| Self::Both((local, *remote)))
-    }
+/// Prepare a default error.
+fn default_err() -> Error {
+    Error::new(ErrorKind::NotFound, "Local interface MTU not found")
```

Comment:
If I understand correctly `default_err()` is often called in cases where we are **very** sure it will never happen, e.g. a `usize` to `u32` conversion.

I understand the rational, treating MTU discovery as optional, thus not worth a panic in production.

That said, how about for the cases we are **very** sure, have an alternative mirror implementation, just with a `debug_assert` to panic in debug mode.

Reply:
Do you mean adding e.g.
```rust
debug_assert!(false, "Local interface name and MTU not found");
```
to the function?

Reply:
How about adding a context aware `debug_assert!` where we are sure the error is impossible.

For example:

``` diff
-        IpAddr::V4(_) => AF_INET.try_into().map_err(|_| default_err())?,
+        IpAddr::V4(_) => AF_INET.try_into().map_err(|_| {debug_assert!(false, "expect AF_INET to fit IPAddr::V4"); default_err()})?,

Reply:
If we think it's impossible that `debug_assert` would never hit? I think it would make the code less readable to do this in many places.